src/osd/OSD.cc

```diff
@@ -10511,6 +10511,16 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
     return;    // OSD shutdown, discard.
   }
 
+  // If no work item was returned, just bail out for now
+  if (!oitem.has_value()) {
+    dout(10) << __func__ << " no work item, returning immediately" << dendl;
+    sdata->shard_lock.unlock();
+    handle_oncommits(oncommits);
+    return;
```

Comment:
What causes us to wake up later when something is allowed to execute?

Reply:
The current logic doesn't result in any sleep. We just return back to the shardedThreadpool_worker loop if the scheduler is unable to return any item. The next dequeue would either result in an item from a client with higher priority being scheduled (if a new higher priority item was enqueued in the meantime) or the existing item on the limit heap if the limit was restored.

The dequeue is attempted only after ensuring that the scheduler queue is not empty.

Reply:
We could end up busy waiting if the scheduler cannot return an item during a particular dequeue operation if for e.g. limits have been exceeded. Therefore, use the “when_ready” response from the scheduler to wait on the condition rather than potentially busy wait in the shardedThreadpool_worker loop.

Reply:
@athanatos Please take a look at the changes to implement the wait using the condition variable. The tests with these changes are showing slightly degraded performance from the client perspective when compared to the last change. This most probably could be due to the time spent for 'double' to 'time_point' conversion.

Reply:
Can you elaborate on the tests?  The double->time_point conversion isn't likely to amount to much overhead, particularly since it should only happen when we wanted the thread to sleep anyway.

Reply:
The tests are the same recovery tests I have been running before. I will run them again with the latest changes and get back.

Reply:
I am running the same recovery test as usual. I will re-test with the latest changes and compare again.

Reply:
waiting and then returning like this requires us to have a 2nd wakeup to execute the next op.

If we structure things as a while loop around dequeuing through waiting, we can immediately process the op if there is one ready to go.

e.g. something like:

```
while (!op_item) {
  is_stopping_check() 
  op_item = dequeue()
  if (wait) {
    wait_until(...)
  }
}

Reply:
@athanatos @jdurgin I have pushed the while loop implementation. Please take a look.